packages/babel-compiler/babel-compiler.js

```diff
@@ -122,46 +126,63 @@ BCp.processOneFileForTarget = function (inputFile, source) {
       },
     };
 
-    this.inferTypeScriptConfig(
-      features, inputFile, cacheOptions.cacheDeps);
+    this.inferTypeScriptConfig(features, inputFile, cacheOptions.cacheDeps);
 
     var babelOptions = Babel.getDefaultOptions(features);
     babelOptions.caller = { name: "meteor", arch };
 
-    this.inferExtraBabelOptions(
-      inputFile,
-      babelOptions,
-      cacheOptions.cacheDeps
-    );
+    this.inferExtraBabelOptions(inputFile, babelOptions, cacheOptions.cacheDeps);
 
     babelOptions.sourceMaps = true;
     babelOptions.filename =
       babelOptions.sourceFileName = packageName
-      ? "packages/" + packageName + "/" + inputFilePath
-      : inputFilePath;
+        ? "packages/" + packageName + "/" + inputFilePath
+        : inputFilePath;
 
     if (this.modifyBabelConfig) {
       this.modifyBabelConfig(babelOptions, inputFile);
     }
 
     try {
       var result = profile('Babel.compile', function () {
+        // Determine if SWC should be used based on package and file criteria.
+        const packagesSkipSwc = [];
+        const fileSkipSwc = ['webapp_server.js']; // top level await
+        const shouldUseSwc = !packagesSkipSwc.includes(packageName) &&
+          !fileSkipSwc.includes(inputFilePath);
+
+        // Check RAM cache
         let compilation;
         try {
-          const packagesSkipSwc = [];
-          const fileSkipSwc = ['webapp_server.js']; // top level await
-
-          // Determine if SWC should be used based on package and file criteria.
-          const shouldUseSwc =
-            !packagesSkipSwc.includes(packageName) &&
-              !fileSkipSwc.includes(inputFilePath);
-
           if (shouldUseSwc) {
-            const isTypescriptSyntax =
-              inputFilePath.endsWith('.ts') || inputFilePath.endsWith('.tsx');
+            // Create a cache key based on the source hash and the compiler used.
+            const cacheKey = toBeAdded.hash;
+            const cacheContext = '.swc-cache';
+
+            // Check RAM cache
+            compilation = self._swcCache[cacheKey];
+            // Check file system cache if enabled
+            if (!compilation && self.cacheDirectory) {
+              const cacheFilePath = path.join(self.cacheDirectory, cacheContext, cacheKey + '.json');
+              if (fs.existsSync(cacheFilePath)) {
+                try {
+                  compilation = JSON.parse(fs.readFileSync(cacheFilePath, 'utf8'));
```

Comment:
What if we use a `require` instead of reading the file and then using JSON.parse? I don't know how it would affect performance, but it would be nice to test this option.

Reply:
Did a benchmark: https://gist.github.com/italojs/00cd47fbd09271031c6ec87bc20cb78b
Require is pretty faster

```
/* Results from my local Intel Mac i9 16gb
fs.readFileSync + JSON.parse:
  Min: 0.0174 ms
  Max: 0.7195 ms
  Average: 0.0221 ms
require:
  Min: 0.0006 ms
  Max: 0.4166 ms
  Average: 0.0010 ms
fs.readFile:
  Min: 0.0701 ms
  Max: 5.3427 ms
  Average: 0.1006 ms
Comparison:
  1. require (Avg: 0.0010 ms)
  2. fs.readFileSync + JSON.parse (Avg: 0.0221 ms)
  3. fs.readFile + JSON.parse (Avg: 0.1006 ms)
*/
```
Maybe we need to take care of relative paths that seem not to work for Windows as described. [here](https://github.com/meteor/meteor/pull/13665#issuecomment-2754637053)

Reply:
it's another opportunity to gain some ms in the bundler since a higher I/O is a huge problem in our bundler today

Reply:
Oh, I have seen the pattern of requires in other caches on Meteor code, so good point @italojs. I appreciate that benchmark you did. I will apply that 🚀

Reply:
`require` catches the result and wouldn't re-read the disk which makes it much faster in the benchmark. The require caching wouldn't have any benefit here since each file is already cached in-memory.

Reply:
I tested `JSON.parse(fs.readFileSync` versus `require` and found the profiles run slower. The cache file is read every time the Meteor app restarts, during each phase shown in the image below. I also highlighted some packages that show additional milliseconds added with that approach, which appears to happen with many of the package read sequentially.

![image](https://github.com/user-attachments/assets/a08d5461-ff74-447c-8896-da12a385cd03)

The benchmark above @italojs was using the Node version Meteor uses? Maybe there was some kind of optimziation on latest one we use.

Reply:
I’m unable to use the standard Node require inside a Meteor package like babel-compiler. Meteor packages use `Npm.require`, which applies its own module resolution logic. The pure Node require doesn’t seem to work in that context.

Reply:
I would suggest two options:
- importing and using the cache system from meteor-babel
- using fs.readfile and json.parse. When used in the cache, I wouldn't expect `require` to be significantly faster.

Reply:
I think sticking with `fs.readFileSync + JSON.parse`. I tried enabling the cache through the `meteor-babel` implementation, but performance actually got worse (see attached). The current approach, which I refactored, runs faster. I suspect that the re-hashing of files for each operation adds significant overhead. In any case, I don’t want to spend more time on this right now, so I’ll leave it open for others to explore further improvements.

![image](https://github.com/user-attachments/assets/1573f2d4-4057-426f-ba7f-beeb59c65ff3)

Reply:
I think sticking with `fs.readFileSync + JSON.parse` makes sense. I tried enabling the cache through the `meteor-babel` implementation, but performance I feel it slightly worse in the several runs I performed, the current approach runs faster. I suspect that the re-hashing of files for each operation adds overhead. In any case, I don’t want to spend more time on this right now.